src/main/java/org/apache/cassandra/sidecar/cdc/CDCReaderService.java

```diff
@@ -0,0 +1,157 @@
+package org.apache.cassandra.sidecar.cdc;
+
+import java.io.InvalidObjectException;
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.Host;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import org.apache.cassandra.schema.KeyspaceMetadata;
+import org.apache.cassandra.schema.Schema;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.sidecar.CQLSession;
+
+/**
+ * Cassandra's real-time change data capture service.
+ */
+@Singleton
+public class CDCReaderService implements Host.StateListener
+{
+    private static final Logger logger = LoggerFactory.getLogger(CDCReaderService.class);
+    private final CDCIndexWatcher cdcIndexWatcher;
+    private final CDCRawDirectoryMonitor cdcRawDirectoryMonitor;
+    private final CQLSession session;
+    private final CassandraConfig cassandraConfig;
+
+    @Inject
+    public CDCReaderService(CDCIndexWatcher cdcIndexWatcher, CDCRawDirectoryMonitor monitor, CQLSession session,
+                            CassandraConfig cassandraConfig)
+    {
+        this.cdcRawDirectoryMonitor = monitor;
+        this.cdcIndexWatcher = cdcIndexWatcher;
+        this.session = session;
+        this.cassandraConfig = cassandraConfig;
+    }
+
+    public synchronized void start()
+    {
+        try
+        {
+            // Wait until the Cassandra server is UP to load configs and start subsystems. There's no guarantee the
+            // config is valid otherwise.
+            waitForCassandraServer();
+            Cluster cluster = session.getLocalCql().getCluster();
+            if (cluster == null)
+            {
+                throw new InvalidObjectException("Cannot connect to the local Cassandra node");
+            }
+
+            // Ensure Cassandra config is valid and remove mutable data paths from the config
+            // to ensure CDC reader doesn't accidentally step on Cassandra data.
+            this.cassandraConfig.init();
+            // TODO : Load metadata from the CQLSession.
+            Schema.instance.loadFromDisk(false);
+            this.cassandraConfig.muteConfigs();
+
+            for (String keySpace : Schema.instance.getKeyspaces())
+            {
+                logger.info("Keyspace : {}", keySpace);
+                KeyspaceMetadata keyspaceMetadata = Schema.instance.getKeyspaceMetadata(keySpace);
+                if (keyspaceMetadata == null)
+                {
+                    continue;
+                }
+                for (TableMetadata tableMetadata : keyspaceMetadata.tablesAndViews())
+                {
```

Comment:
here ,I think we can add a loop , that every time we check if there is a table have got its cdc enable in schema info , 
if all table 's cdc is not enabled ,we can just go to sleep  for a while , such as commitlog flush period ,in this way for 
users that do not use cdc ,we can just save some resource. 
also every loop the schema should refresh (I think the cassandra driver have done ,we just use it );
when a table is cdc enabled ,we go to next step;m
my code here is :

// Ensure Cassandra config is valid and remove mutable data paths from the config
            // to ensure CDC reader doesn't accidentally step on Cassandra data.
            this.cassandraConfig.init();

            while (true)
            {
                int cdcEnableTables = 0;
                Metadata metadata = this.session.getLocalCql().getCluster().getMetadata();
                List<KeyspaceMetadata> keyspaceMetadatas = metadata.getKeyspaces();
                for (KeyspaceMetadata keyspaceMetadata : keyspaceMetadatas)
                {
                    if (keyspaceMetadata == null)
                    {
                        continue;
                    }
                    for (TableMetadata tableMetadata : keyspaceMetadata.getTables())
                    {
                        if (tableMetadata.getOptions().isCDC())
                            ++cdcEnableTables;
                    }
                }
                if (cdcEnableTables != 0)
                    break;
                logger.warn("There is no table enable the cdc , just sleep for %", DatabaseDescriptor.getCommitLogSyncPeriod());
                Thread.sleep(DatabaseDescriptor.getCommitLogSyncPeriod());
            }

            this.cassandraConfig.muteConfigs();
            // Start monitoring the cdc_raw directory
            this.cdcRawDirectoryMonitor.startMonitoring();

Reply:
This code was just added for debugging. Users can alter tables to enable CDC at any time. Sidecar shouldn't wait. Also, whatever user does is not visible to us without reloading metadata. This is where the `CDCSchemaChangeListener` is helpful.
+1 for the suggestion. When `CDCSchemaChangeListener` is working, we can add a guard rail to `CDCIndexWatcher` so it won't read commit log entries when there are no tables with CDC enabled.